### PR TITLE
unique_name and unique_hidden_name

### PR DESCRIPTION
**Pull request by ghost**
### Description of work
Create helper functions to create unique workspace names to avoid collisions with other workspace names. This could be useful for testing, temporary workspace or people who lack the creativity to name their workspaces. 

#### Summary of work

- `unique_name` creates a unique workspace name of *n* number of random characters. This function also allows for a prefix or suffix to be specified to help with grouping workspaces. 
- `unique_hidden_name` creates a unique workspace name that has a prefix of '__' (double under-bar). If this name is used to create a workspace, the workspace will not be visible with calls like `mtd.getObjectNames()` but can be accessed using `mtd[hidden_workspace_name]` or `mtd.retrieve(hidden_workspace_name)`

Functions were created in `AnalysisDataService` and are accessible in python as:
- `mtd.unique_name` or `AnalysisDataService.unique_name`
- `mtd.unique_hidden_name` or `AnalysisDataService.unique_hidden_name`

References #36387 and associated with [EWM #2940](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=2940)

### To test:
Unit tests have been included for C++ and Python to show functionality. The following command will run tests in both languages:
```
ctest -R AnalysisDataServiceTest
```

Manual Test
```python
from mantid.simpleapi import mtd

wsname = mtd.unique_name()
CreateSampleWorkspace(OutputWorkspace=wsname)
assert([wsname] == mtd.getObjectNames()) # workspace created with unique name

hidden_name = mtd.unique_hidden_name()
CreateSampleWorkspace(OutputWorkspace=hidden_name)
assert([wsname] == mtd.getObjectNames()) # workspace created but not visible
assert(mtd.retrieve(hidden_name)) # workspace still accessible with name. 
```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/API/src/AnalysisDataService.cpp [hunk 1]
```diff
@@ -167,6 +169,54 @@ void AnalysisDataServiceImpl::remove(const std::string &name) {
...
+    for (int i = 0; i < n; ++i) {
+      name += getRandomLowercaseLetter();
+    }
+    return name;
+  };
+
+  std::string wsName = "";
+  do {
+    wsName = randomNameGenerator();
+    std::cout << "wsname: " << wsName << "\n";
```
Comment (jmborr) [inline Framework/API/src/AnalysisDataService.cpp:208]: can you remove this printout?

File: Framework/API/src/AnalysisDataService.cpp [hunk 2]
```diff
@@ -167,6 +169,54 @@ void AnalysisDataServiceImpl::remove(const std::string &name) {
...
+    std::string name;
+    for (int i = 0; i < n; ++i) {
+      name += getRandomLowercaseLetter();
+    }
+    return name;
+  };
+
+  std::string wsName = "";
+  do {
+    wsName = randomNameGenerator();
```
Comment (jmborr) [inline Framework/API/src/AnalysisDataService.cpp:207]: what should be compared against the recorded names is `wsName = prefix + randomNameGenerator() + suffix`

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp [hunk 3]
```diff
@@ -85,5 +85,10 @@ void export_AnalysisDataService() {
       .def("addToGroup", &AnalysisDataServiceImpl::addToGroup, (arg("groupName"), arg("wsName")),
            "Add a workspace in the ADS to a group in the ADS")
       .def("removeFromGroup", &AnalysisDataServiceImpl::removeFromGroup, (arg("groupName"), arg("wsName")),
-           "Remove a workspace from a group in the ADS");
+           "Remove a workspace from a group in the ADS")
+      .def("unique_name", &AnalysisDataServiceImpl::uniqueName,
+           (arg("self"), arg("n") = 5, arg("prefx") = "", arg("suffix") = ""),
+           "Return a randomly generated unique name for a workspace.")
```
Comment (jmborr) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp:91]:
```python
In [2]: mtd.unique_name?
Docstring:
unique_name( (AnalysisDataServiceImpl)self [, (int)n=5 [, (str)prefx='' [, (str)suffix='']]]) -> str :
    Return a randomly generated unique name for a workspace.
Type:      method
```
the docstring doesn't explain the arguments

**Review by jmborr — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp [hunk 4]
```diff
@@ -85,5 +85,14 @@ void export_AnalysisDataService() {
...
       .def("removeFromGroup", &AnalysisDataServiceImpl::removeFromGroup, (arg("groupName"), arg("wsName")),
-           "Remove a workspace from a group in the ADS");
+           "Remove a workspace from a group in the ADS")
+      .def("unique_name", &AnalysisDataServiceImpl::uniqueName,
+           (arg("self"), arg("n") = 5, arg("prefix") = "", arg("suffix") = ""),
+           ""
+           "Return a randomly generated unique name for a workspace \
+           `n` : length of string of random numbers, `prefix` : String to be prepended to the generated string, \
+           `suffix` : String to be appended to the generated string, `returns` : (prefix + n*random characters + suffix) "
+           "")
```
Comment (jmborr) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp:95]:
```suggestion
      .def("unique_name", &AnalysisDataServiceImpl::uniqueName,
           (arg("self"), arg("n") = 5, arg("prefix") = "", arg("suffix") = ""),
           "Return a randomly generated unique name for a workspace\n"
           "\nParameters\n"
           "----------\n"
           "n: str\n"
           "   length of string of random numbers\n"
           "prefix: str\n"
           "   String to be prepended to the generated string\n"
           "suffix: str\n"
           "   String to be appended to the generated string\n"
           "\nReturns\n"
           "-------\n"
           "str\n"
           "    prefix + n*random characters + suffix\n"
           )
```
This is how it looks:
Capture

**Review by jmborr — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp (hunk 4 above)
Comment (searscr) [reply Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp:95]: I tried doing it that way. I'm not sure how you are getting the docstring to print. I was building `docs-html` and checking what it looked like on the `AnalysisDataServiceImpl` page and if you do it the way you suggested it jumbles it all up.

**Review by jmborr — APPROVED** [review]

**Pull request merged**